File: src/main/java/io/github/intellij/dlanguage/codeinsight/dcd/completions/DCDModel.java
package io.github.intellij.dlanguage.codeinsight.dcd.completions;

import org.jetbrains.annotations.Nullable;

public class DCDModel {
    public DCDCompletionKind kind;
    public String value;

    @Nullable
    public String typeText = null;

    public DCDModel(String name, String id, @Nullable String type) {
        value = name;
        kind = fromId(id);
        typeText = type;
    }

    public DCDModel(String name, String id) {
        this(name, id, null);
    }

    public DCDCompletionKind fromId(String id) {
        switch (id) {
            case "c":
                return DCDCompletionKind.ClassName;
            case "i":
                return DCDCompletionKind.InterfaceName;
            case "s":
                return DCDCompletionKind.StructName;
            case "u":
                return DCDCompletionKind.UnionName;
            case "v":
                return DCDCompletionKind.VariableName;
            case "m":
                return DCDCompletionKind.MemberVariableName;
            case "k":
                return DCDCompletionKind.Keyword;
            case "f":
                return DCDCompletionKind.Function;
            case "g":
                return DCDCompletionKind.EnumName;
            case "e":
                return DCDCompletionKind.EnumMember;
            case "P":
                return DCDCompletionKind.PackageName;
            case "M":
                return DCDCompletionKind.ModuleName;
            case "a":
                return DCDCompletionKind.Array;
            case "A":
                return DCDCompletionKind.AssociativeArray;
            case "l":
                return DCDCompletionKind.AliasName;
            case "t":
                return DCDCompletionKind.TemplateName;
            case "T":
                return DCDCompletionKind.MixinTemplate;
        }
        throw new RuntimeException("Unknow DCD Completion kind: " + id);
    }
}

File: src/main/java/io/github/intellij/dlanguage/codeinsight/dcd/completions/DCDModel.kt
package io.github.intellij.dlanguage.codeinsight.dcd.completions

class DCDModel(var value: String, id: String, type: String? = null)
{
    var kind: DCDCompletionKind

    var typeText: String? = null

    init
    {
        kind = fromId(id)
        typeText = type

    }

    fun fromId(id: String): DCDCompletionKind
    {
        when (id)
        {
            "c" -> return DCDCompletionKind.ClassName
            "i" -> return DCDCompletionKind.InterfaceName
            "s" -> return DCDCompletionKind.StructName
            "u" -> return DCDCompletionKind.UnionName
            "v" -> return DCDCompletionKind.VariableName
            "m" -> return DCDCompletionKind.MemberVariableName
            "k" -> return DCDCompletionKind.Keyword
            "f" -> return DCDCompletionKind.Function
            "g" -> return DCDCompletionKind.EnumName
            "e" -> return DCDCompletionKind.EnumMember
            "P" -> return DCDCompletionKind.PackageName
            "M" -> return DCDCompletionKind.ModuleName
            "a" -> return DCDCompletionKind.Array
            "A" -> return DCDCompletionKind.AssociativeArray
            "l" -> return DCDCompletionKind.AliasName
            "t" -> return DCDCompletionKind.TemplateName
            "T" -> return DCDCompletionKind.MixinTemplate
        }
        throw RuntimeException("Unknow DCD Completion kind: $id")
    }
}
